torch/cuda/__init__.py

```diff
@@ -322,40 +323,83 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
 
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
 
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
 
     Args:
-        stream (Stream): selected stream. This manager is a no-op if it's
+        Stream (Stream): selected stream. This manager is a no-op if it's
             ``None``.
+    .. note:: Streams are per-device.
+    """
+    cur_stream : Optional['torch.cuda.Stream']
+
+    def __init__(self, stream: Optional['torch.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        # Initialize the below streams to default stream on the current device
+        self.src_prev_stream = torch.cuda.default_stream(None)
+        self.dst_prev_stream = torch.cuda.default_stream(None)
+
+    def __enter__(self):
+        self.idx = _get_device_index(device=None, optional=True)
+        # If there is no CUDA device available, return
+        if self.idx == -1:
+            return
```

Comment:
This index variable is required to check if there is no CUDA device is available. Also, it is later referred in `__exit__` to check the same. If we want to eliminate `self.idx` completely, then we would need to call this API `_get_device_index` in `__enter__` and `__exit__`

Reply:
The old implementation of this context manager didn't have this check. It only checks if `stream` is `None` and exits early in that case.

Reply:
Yes, it was added in case the CUDA device isn't available.

Reply:
How come the old implementation did not have code like this? Is it a bug in their implementation?

If there is no CUDA device available, the stream must be `None`, right?

Reply:
How come the old implementation did not have code like this? Is it a bug in their implementation?
>> Maybe.Because if the CUDA device becomes unavailable inside `__enter__` , then `stream` is still not `None`(as it was set in `__init__`  and we don't return.

Reply:
Ok, so it is an issue introduced by this implementation because `__init__` and `__enter__` are two separate functions. I think I understand now.

But under what circumstances could a device be available in `__init__` and not in `__enter__`?

Reply:
The only scenario that I can think of is when the connection to device is lost in `__enter__`. Although the window is very tiny, there are chances of this still happening.

Reply:
By this reasoning, this code could also be wrong, because you could test this in line 347, and then immediately afterwards we could lose connection to device.

The more realistic approach is to assume (like most code in PyTorch does) that the set of available CUDA devices stays fixed for the entirety of the PyTorch process. Then this code is unnecessary

Reply:
Sounds good. Fixed now